File: packages/b2c-tooling-sdk/src/config/dw-json.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * dw.json configuration file loading.
 *
 * This module provides utilities for loading B2C Commerce configuration from
 * dw.json files, the standard configuration format used by B2C development tools.
 *
 * @module config
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import type {AuthMethod} from '../auth/types.js';
import {getLogger} from '../logging/logger.js';
import {normalizeConfigKeys} from './mapping.js';

/**
 * Configuration structure for dw.json after key normalization.
 *
 * All keys are normalized to camelCase by `normalizeConfigKeys()` when loading.
 * Both camelCase and kebab-case are accepted in the raw file; the interface
 * documents the canonical (post-normalization) field names.
 *
 * Legacy aliases (e.g., `server`, `secureHostname`, `passphrase`, `selfsigned`,
 * `cloudOrigin`, `scapi-shortcode`) are also accepted and mapped to their
 * canonical names during normalization.
 */
export interface DwJsonConfig {
  /** Instance name (for multi-config files) */
  name?: string;
  /** Whether this config is active (for multi-config files) */
  active?: boolean;
  /** B2C instance hostname */
  hostname?: string;
  /** Code version for deployments */
  codeVersion?: string;
  /** Username for Basic auth (WebDAV) */
  username?: string;
  /** Password/access-key for Basic auth (WebDAV) */
  password?: string;
  /** OAuth client ID */
  clientId?: string;
  /** OAuth client secret */
  clientSecret?: string;
  /** OAuth scopes */
  oauthScopes?: string[];
  /** SCAPI short code */
  shortCode?: string;
  /** Alternate hostname for WebDAV (if different from main hostname) */
  webdavHostname?: string;
  /** Allowed authentication methods in priority order */
  authMethods?: AuthMethod[];
  /** Account Manager hostname for OAuth */
  accountManagerHost?: string;
  /** MRT project slug */
  mrtProject?: string;
  /** MRT environment name (e.g., staging, production) */
  mrtEnvironment?: string;
  /** MRT API key */
  mrtApiKey?: string;
  /** MRT cloud origin URL */
  mrtOrigin?: string;
  /** Tenant/Organization ID for SCAPI */
  tenantId?: string;
  /** ODS API hostname */
  sandboxApiHost?: string;
  /** Default content library ID for content export/list commands */
  contentLibrary?: string;
  /** Path to PKCS12 certificate file for mTLS (two-factor auth) */
  certificate?: string;
  /** Passphrase for the certificate */
  certificatePassphrase?: string;
  /** Whether to skip SSL/TLS certificate verification (self-signed certs) */
  selfSigned?: boolean;
}

/**
 * dw.json with multi-config support (configs array).
 */
export interface DwJsonMultiConfig extends DwJsonConfig {
  /** Array of named instance configurations */
  configs?: DwJsonConfig[];
}

/**
 * Options for loading dw.json.
 */
export interface LoadDwJsonOptions {
  /** Named instance to select from configs array */
  instance?: string;
  /** Explicit path to dw.json (skips searching if provided) */
  path?: string;
  /** Starting directory for search (defaults to cwd) */
  startDir?: string;
}

/**
 * Result of loading dw.json configuration.
 */
export interface LoadDwJsonResult {
  /** The parsed configuration */
  config: DwJsonConfig;
  /** The path to the dw.json file that was loaded */
  path: string;
}

/**
 * Finds dw.json by searching upward from the starting directory.
 *
 * @param startDir - Directory to start searching from (defaults to cwd)
 * @returns Path to dw.json if found, undefined otherwise
 *
 * @example
 * const dwPath = findDwJson();
 * if (dwPath) {
 *   console.log(`Found dw.json at ${dwPath}`);
 * }
 */
export function findDwJson(startDir: string = process.cwd()): string | undefined {
  let dir = startDir;
  const root = path.parse(dir).root;

  while (dir !== root) {
    const dwJsonPath = path.join(dir, 'dw.json');
    if (fs.existsSync(dwJsonPath)) {
      return dwJsonPath;
    }
    dir = path.dirname(dir);
  }

  return undefined;
}

/**
 * Selects the appropriate config from a multi-config dw.json.
 *
 * Selection priority:
 * 1. Named instance (if `instance` option provided)
 * 2. Config marked as `active: true`
 * 3. Root-level config
 */
function selectConfig(json: DwJsonMultiConfig, instanceName?: string): DwJsonConfig | undefined {
  const logger = getLogger();

  // Single config or no configs array
  if (!Array.isArray(json.configs) || json.configs.length === 0) {
    logger.trace(
      {selection: 'root', instanceName: json.name},
      `[DwJsonSource] Selected config "${json.name ?? 'root'}" (single config)`,
    );
    return json;
  }

  // Find by instance name
  if (instanceName) {
    // Check root first
    if (json.name === instanceName) {
      logger.trace(
        {selection: 'named', instanceName},
        `[DwJsonSource] Selected config "${instanceName}" by name (root)`,
      );
      return json;
    }
    // Then check configs array
    const found = json.configs.find((c) => c.name === instanceName);
    if (found) {
      logger.trace({selection: 'named', instanceName}, `[DwJsonSource] Selected config "${instanceName}" by name`);
      return found;
    }
    // Instance explicitly requested but not found - return undefined
    logger.trace({requestedInstance: instanceName}, `[DwJsonSource] Named instance "${instanceName}" not found`);
    return undefined;
  }

  // Find active config
  if (json.active === false) {
    // Root is inactive, look for active in configs
    const activeConfig = json.configs.find((c) => c.active === true);
    if (activeConfig) {
      logger.trace(
        {selection: 'active', instanceName: activeConfig.name},
        `[DwJsonSource] Selected config "${activeConfig.name}" by active flag`,
      );
      return activeConfig;
    }
  }

  // Default to root config
  logger.trace(
    {selection: 'root', instanceName: json.name},
    `[DwJsonSource] Selected config "${json.name ?? 'root'}" (default to root)`,
  );
  return json;
}

/**
 * Loads configuration from a dw.json file.
 *
 * If an explicit path is provided, uses that file. Otherwise, looks for dw.json
 * in the startDir (or cwd). Does NOT search parent directories.
 *
 * Use `findDwJson()` if you need to search upward through parent directories.
 *
 * @param options - Loading options
 * @returns The parsed config with its path, or undefined if no dw.json found
 *
 * @example
 * // Load from ./dw.json (current directory)
 * const result = loadDwJson();
 * if (result) {
 *   console.log(`Loaded from ${result.path}`);
 *   console.log(result.config.hostname);
 * }
 *
 * // Load from specific directory
 * const result = loadDwJson({ startDir: '/path/to/project' });
 *
 * // Use named instance
 * const result = loadDwJson({ instance: 'staging' });
 *
 * // Explicit path
 * const result = loadDwJson({ path: './config/dw.json' });
 */
export function loadDwJson(options: LoadDwJsonOptions = {}): LoadDwJsonResult | undefined {
  const logger = getLogger();

  // If explicit path provided, use it. Otherwise default to ./dw.json (no upward search)
  const dwJsonPath = options.path ?? path.join(options.startDir || process.cwd(), 'dw.json');

  logger.trace({path: dwJsonPath}, '[DwJsonSource] Checking for config file');

  if (!fs.existsSync(dwJsonPath)) {
    logger.trace({path: dwJsonPath}, '[DwJsonSource] No config file found');
    return undefined;
  }

  try {
    const content = fs.readFileSync(dwJsonPath, 'utf8');
    const raw = JSON.parse(content) as Record<string, unknown>;

    // Normalize root-level keys to camelCase
    const normalized = normalizeConfigKeys(raw) as DwJsonMultiConfig;

    // Normalize keys in each configs[] item
    if (Array.isArray(normalized.configs)) {
      normalized.configs = normalized.configs.map(
        (item) => normalizeConfigKeys(item as Record<string, unknown>) as DwJsonConfig,
      );
    }

    const config = selectConfig(normalized, options.instance);
    if (!config) {
      return undefined;
    }
    return {
      config,
      path: dwJsonPath,
    };
  } catch (error) {
    // Invalid JSON or read error - log at trace level and re-throw
    // The resolver will catch this and create a SOURCE_ERROR warning
    const message = error instanceof Error ? error.message : String(error);
    logger.trace({path: dwJsonPath, error: message}, '[DwJsonSource] Failed to parse config file');
    throw error;
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/config/dw_json.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""dw.json configuration file loading.

Mirrors ``src/config/dw-json.ts``. Loads B2C Commerce configuration from
``dw.json`` files, the standard format used by B2C development tools.

The TypeScript functions are async (``fs/promises``); config file access is not
a hot path, so this port uses synchronous file IO. Files are written with the
same 2-space indent + trailing newline the TypeScript SDK produces, so a
``dw.json`` edited from Python stays interoperable.

``DwJsonConfig`` / ``DwJsonMultiConfig`` are plain ``dict[str, Any]`` with
camelCase keys (post :func:`normalize_config_keys`), matching the on-disk format.
"""

from __future__ import annotations

import json as json_module
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from b2c_tooling_sdk.config.mapping import normalize_config_keys
from b2c_tooling_sdk.logging import get_logger

#: A single dw.json instance config (camelCase keys, post key-normalization).
DwJsonConfig = dict[str, Any]
#: A dw.json file that may carry a ``configs`` array of named instances.
DwJsonMultiConfig = dict[str, Any]


@dataclass
class LoadDwJsonResult:
    """The selected dw.json config plus the file path it came from."""

    config: DwJsonConfig
    path: str


@dataclass
class LoadFullDwJsonResult:
    """The full (multi-config) dw.json structure plus its file path."""

    config: DwJsonMultiConfig
    path: str


def _default_dw_json_path(
    *,
    path: str | None,
    project_directory: str | None = None,
    working_directory: str | None = None,
) -> str:
    """Resolve the dw.json path: explicit path, else ``<project|working|cwd>/dw.json``."""
    if path is not None:
        return path
    base = project_directory or working_directory or os.getcwd()
    return os.path.join(base, "dw.json")


def find_dw_json(project_directory: str | None = None) -> str | None:
    """Find dw.json by searching upward from ``project_directory`` (defaults to cwd)."""
    start = project_directory or os.getcwd()
    directory = Path(start).resolve()
    root = Path(directory.anchor)

    current = directory
    while current != root:
        candidate = current / "dw.json"
        if candidate.exists():
            return str(candidate)
        current = current.parent
    # Check the filesystem root itself for parity with a top-level dw.json.
    candidate = root / "dw.json"
    return str(candidate) if candidate.exists() else None


def _select_config(json_data: DwJsonMultiConfig, instance_name: str | None = None) -> DwJsonConfig | None:
    """Select a config from a multi-config dw.json.

    Priority: named instance (if requested) -> ``active: true`` -> root config.
    """
    logger = get_logger("config.dw_json")
    configs = json_data.get("configs")

    if not isinstance(configs, list) or len(configs) == 0:
        logger.debug("[DwJsonSource] Selected config %r (single config)", json_data.get("name") or "root")
        return json_data

    entries: list[DwJsonConfig] = configs

    if instance_name:
        if json_data.get("name") == instance_name:
            logger.debug("[DwJsonSource] Selected config %r by name (root)", instance_name)
            return json_data
        for c in entries:
            if c.get("name") == instance_name:
                logger.debug("[DwJsonSource] Selected config %r by name", instance_name)
                return c
        logger.debug("[DwJsonSource] Named instance %r not found", instance_name)
        return None

    for c in entries:
        if c.get("active") is True:
            logger.debug("[DwJsonSource] Selected config %r by active flag", c.get("name"))
            return c

    logger.debug("[DwJsonSource] Selected config %r (default to root)", json_data.get("name") or "root")
    return json_data


def load_full_dw_json(
    *,
    path: str | None = None,
    project_directory: str | None = None,
    working_directory: str | None = None,
) -> LoadFullDwJsonResult | None:
    """Load the raw multi-config dw.json without selecting an instance.

    Returns ``None`` if the file does not exist. Raises on invalid JSON.
    """
    logger = get_logger("config.dw_json")
    dw_json_path = _default_dw_json_path(
        path=path, project_directory=project_directory, working_directory=working_directory
    )

    if not os.path.exists(dw_json_path):
        logger.debug("[DwJsonSource] No config file found: %s", dw_json_path)
        return None

    try:
        with open(dw_json_path, encoding="utf-8") as fh:
            data = json_module.load(fh)
    except Exception as error:
        logger.debug("[DwJsonSource] Failed to parse config file %s: %s", dw_json_path, error)
        raise
    return LoadFullDwJsonResult(config=data, path=dw_json_path)


def save_dw_json(config: DwJsonMultiConfig, file_path: str) -> None:
    """Save a dw.json to disk (2-space indent + trailing newline)."""
    content = json_module.dumps(config, indent=2) + "\n"
    with open(file_path, "w", encoding="utf-8") as fh:
        fh.write(content)


def add_instance(
    instance: DwJsonConfig,
    *,
    path: str | None = None,
    project_directory: str | None = None,
    working_directory: str | None = None,
    set_active: bool = False,
) -> None:
    """Add a new named instance to dw.json, creating the file if needed.

    :raises ValueError: if the instance has no name or the name already exists.
    """
    dw_json_path = _default_dw_json_path(
        path=path, project_directory=project_directory, working_directory=working_directory
    )

    existing: DwJsonMultiConfig = {}
    if os.path.exists(dw_json_path):
        with open(dw_json_path, encoding="utf-8") as fh:
            existing = json_module.load(fh)

    instance_name = instance.get("name")
    if not instance_name:
        raise ValueError("Instance must have a name")

    if existing.get("name") == instance_name:
        raise ValueError(f'Instance "{instance_name}" already exists')

    configs = existing.get("configs")
    if isinstance(configs, list) and any(c.get("name") == instance_name for c in configs):
        raise ValueError(f'Instance "{instance_name}" already exists')

    if set_active:
        instance["active"] = True
        if existing.get("active") is not None:
            existing["active"] = False
        if isinstance(existing.get("configs"), list):
            for c in existing["configs"]:
                if c.get("active") is not None:
                    c["active"] = False

    if not isinstance(existing.get("configs"), list):
        existing["configs"] = []
    existing["configs"].append(instance)

    save_dw_json(existing, dw_json_path)


def remove_instance(
    name: str,
    *,
    path: str | None = None,
    project_directory: str | None = None,
    working_directory: str | None = None,
) -> None:
    """Remove a named instance from dw.json.

    :raises FileNotFoundError: if no dw.json exists.
    :raises ValueError: when removing the root instance or a missing instance.
    """
    dw_json_path = _default_dw_json_path(
        path=path, project_directory=project_directory, working_directory=working_directory
    )
    if not os.path.exists(dw_json_path):
        raise FileNotFoundError("No dw.json file found")

    with open(dw_json_path, encoding="utf-8") as fh:
        existing = json_module.load(fh)

    if existing.get("name") == name:
        raise ValueError(f'Cannot remove root instance "{name}". Edit dw.json manually to remove root config.')

    configs = existing.get("configs")
    if not isinstance(configs, list) or not any(c.get("name") == name for c in configs):
        raise ValueError(f'Instance "{name}" not found')

    existing["configs"] = [c for c in configs if c.get("name") != name]
    save_dw_json(existing, dw_json_path)


def set_active_instance(
    name: str,
    *,
    path: str | None = None,
    project_directory: str | None = None,
    working_directory: str | None = None,
) -> None:
    """Set a named instance as the active default.

    :raises FileNotFoundError: if no dw.json exists.
    :raises ValueError: if the instance is not found.
    """
    dw_json_path = _default_dw_json_path(
        path=path, project_directory=project_directory, working_directory=working_directory
    )
    if not os.path.exists(dw_json_path):
        raise FileNotFoundError("No dw.json file found")

    with open(dw_json_path, encoding="utf-8") as fh:
        existing = json_module.load(fh)

    found = False
    if existing.get("name") == name:
        found = True
        existing["active"] = True
    elif existing.get("active") is not None:
        existing["active"] = False

    if isinstance(existing.get("configs"), list):
        for c in existing["configs"]:
            if c.get("name") == name:
                found = True
                c["active"] = True
            elif c.get("active") is not None:
                c["active"] = False

    if not found:
        raise ValueError(f'Instance "{name}" not found')

    save_dw_json(existing, dw_json_path)


def load_dw_json(
    *,
    instance: str | None = None,
    path: str | None = None,
    project_directory: str | None = None,
    working_directory: str | None = None,
) -> LoadDwJsonResult | None:
    """Load configuration from a dw.json file (no upward directory search).

    Keys are normalized to camelCase and the appropriate instance is selected.
    Returns ``None`` if no file is found or the named instance is absent. Raises
    on invalid JSON.
    """
    logger = get_logger("config.dw_json")
    dw_json_path = _default_dw_json_path(
        path=path, project_directory=project_directory, working_directory=working_directory
    )

    if not os.path.exists(dw_json_path):
        logger.debug("[DwJsonSource] No config file found: %s", dw_json_path)
        return None

    try:
        with open(dw_json_path, encoding="utf-8") as fh:
            raw = json_module.load(fh)
    except Exception as error:
        logger.debug("[DwJsonSource] Failed to parse config file %s: %s", dw_json_path, error)
        raise

    normalized = normalize_config_keys(raw)
    configs = normalized.get("configs")
    if isinstance(configs, list):
        normalized["configs"] = [normalize_config_keys(item) for item in configs]

    config = _select_config(normalized, instance)
    if config is None:
        return None
    return LoadDwJsonResult(config=config, path=dw_json_path)


__all__ = [
    "DwJsonConfig",
    "DwJsonMultiConfig",
    "LoadDwJsonResult",
    "LoadFullDwJsonResult",
    "add_instance",
    "find_dw_json",
    "load_dw_json",
    "load_full_dw_json",
    "remove_instance",
    "save_dw_json",
    "set_active_instance",
]
